Zotero/Controllers/Sync/SyncController.swift

```diff
@@ -2125,91 +2125,96 @@ final class SyncController: SynchronizationController {
-                self.handleUnchangedFailure(lastVersion: version, libraryId: libraryId, additionalAction: additionalAction)
+                handleUnchangedFailure(lastVersion: version, libraryId: libraryId, additionalAction: additionalAction)
             } else {
                 guard additionalAction?() ?? true else { return }
                 self.processNextAction()
             }
 
         case .quotaLimit(let libraryId):
-            self.removeRemainingUploads(for: libraryId, andAppendError: error, additionalAction: additionalAction)
+            handleQuotaLimit(for: libraryId, andAppendError: error, additionalAction: additionalAction)
 
         default:
             appendAndContinue()
         }
-    }
 
-    /// Handle fatal upload errors for given library. If backend returns code 403 or 413 it means that the user can't upload more files to backend, so no further uploads should be attempted.
-    /// - parameter libraryId: Library identifier which reached quota limit.
-    /// - parameter error: NonFatal error that caused this
-    /// - additiionalAction: Additional actions that should be applied to queue
-    private func removeRemainingUploads(for libraryId: LibraryIdentifier, andAppendError error: SyncError.NonFatal, additionalAction: (() -> Bool)?) {
-        DDLogInfo("Sync: received quota limit for \(libraryId)")
+        /// Handle quota limit error. It means that the user can't upload more files to backend, so no further uploads should be attempted. Should enqueue check for remote changes.
+        /// - parameter libraryId: Library identifier which reached quota limit.
+        /// - parameter error: NonFatal error that caused this
+        /// - additiionalAction: Additional actions that should be applied to queue
+        func handleQuotaLimit(for libraryId: LibraryIdentifier, andAppendError error: SyncError.NonFatal, additionalAction: (() -> Bool)?) {
+            DDLogInfo("Sync: received quota limit for \(libraryId)")
 
-        self.queue.removeAll(where: { action in
-            switch action {
-            case .uploadAttachment(let attachment):
-                return attachment.libraryId == libraryId
+            // Remove all other upload actions from queue, our quota is full, so they'll just fail
+            self.queue.removeAll(where: { action in
+                switch action {
+                case .uploadAttachment(let attachment):
+                    return attachment.libraryId == libraryId
 
-            default:
-                return false
-            }
-        })
+                default:
+                    return false
+                }
+            })
 
-        if !self.nonFatalErrors.contains(error) {
-            self.nonFatalErrors.append(error)
-        }
+            // Uploads failed, we need to check for remote changes for this library, in case there were no previous writes.
+            self.queue.insert(.createLibraryActions(.specific([libraryId]), .onlyDownloads), at: 0)
```

Comment:
Since there is already such an action in the queue for all libraries, with `automatic` options, why is this needed?

Reply:
Got it. 
As I was checking a relevant log, there was this sequence:
```
Sync: createLibraryActions(Zotero.SyncController.Libraries.all, Zotero.SyncController.CreateLibraryActionsOptions.automatic)

Sync: createUploadActions(libraryId: Zotero.LibraryIdentifier.custom(Zotero.RCustomLibraryType.myLibrary), hadOtherWriteActions: false, canEditFiles: true)
```

before the single upload action failed, so I assumed that the first creation would have already enqueued any download actions, or would at least do so after the non fatal error (if the queue order was jumped).